src/main/resources/templates/db_scripts/transactions.sql

```diff
@@ -0,0 +1,97 @@
+
+    toBudget                jsonb;
+    toBudgetAllocated       decimal;
+    toBudgetAvailable       decimal;
+    toBudgetUnavailable     decimal;
+
+    toLedger                jsonb;
+    toLedgerAllocated       decimal;
+    toLedgerAvailable       decimal;
+    toLedgerUnavailable     decimal;
+
+    newBudgetValues         text[];
+    newLedgerValues         text[];
+    amount                  decimal;
+
+    BEGIN
+      amount = (NEW.jsonb->>'amount')::decimal;
+
+      -- Recalculate Allocation
+      IF (NEW.jsonb->>'transactionType' = 'Allocation') THEN
+        --check if fromFundId exists
+        IF (NEW.jsonb->'fromFundId' IS NOT NULL) THEN
+          -- Update Budget identified by the transactions fiscal year (fiscalYearId) and source fund (fromFundId)
+          SELECT INTO fromBudget (jsonb::jsonb) FROM  ${myuniversity}_${mymodule}.budget WHERE (jsonb->>'fiscalYearId' = NEW.jsonb->>'fiscalYearId' AND jsonb->>'fundId' = NEW.jsonb->>'fromFundId');
+
+          fromBudgetAllocated = (SELECT COALESCE(fromBudget->>'allocated', '0'))::decimal - amount;
+          fromBudgetAvailable = (SELECT COALESCE(fromBudget->>'available', '0'))::decimal - amount;
+          fromBudgetUnavailable = (SELECT COALESCE(fromBudget->>'unavailable', '0'))::decimal + amount;
+          newBudgetValues = '{allocated,' || fromBudgetAllocated || ', available, ' || fromBudgetAvailable || ', unavailable, ' || fromBudgetUnavailable ||'}';
+
+          UPDATE ${myuniversity}_${mymodule}.budget SET jsonb = jsonb || json_object(newBudgetValues)::jsonb WHERE (jsonb->>'fiscalYearId' = NEW.jsonb->>'fiscalYearId' AND jsonb->>'fundId' = NEW.jsonb->>'fromFundId');
+
+
+          -- Update Ledger identified by the transaction's fiscal year (fiscalYearId) and the source fund (fromFundId)
+          SELECT INTO fromLedger (jsonb::jsonb) FROM  ${myuniversity}_${mymodule}.ledger  WHERE (jsonb->>'id' = (SELECT jsonb->>'ledgerId' FROM ${myuniversity}_${mymodule}.fund WHERE (jsonb->>'id' = fromBudget->>'fundId')));
+
+          fromLedgerAllocated = (SELECT COALESCE(fromLedger->>'allocated', '0'))::decimal + amount;
+          fromLedgerAvailable = (SELECT COALESCE(fromLedger->>'available', '0'))::decimal - amount;
+          fromLedgerUnavailable = (SELECT COALESCE(fromLedger->>'unavailable', '0'))::decimal + amount;
+          newLedgerValues = '{allocated,' || fromLedgerAllocated || ', available, ' || fromLedgerAvailable || ', unavailable, ' || fromLedgerUnavailable ||'}';
+
+          UPDATE ${myuniversity}_${mymodule}.ledger SET jsonb = jsonb || json_object(newLedgerValues)::jsonb WHERE (jsonb->>'id' = fromLedger->>'id');
+
+        END IF;
+        -- Update Budget identified by the transaction's fiscal year (fiscalYearId) and the destination fund (toFundId)
+          SELECT INTO toBudget (jsonb::jsonb) FROM  ${myuniversity}_${mymodule}.budget WHERE (jsonb->>'fiscalYearId' = NEW.jsonb->>'fiscalYearId' AND jsonb->>'fundId' = NEW.jsonb->>'toFundId');
```

Comment:
`toFundId` field is not required, so it is possible that it will be empty. @folio-org/acquisitions  Where do you think it would be better to check for the presence of this field in storage-module or bl-module?

Reply:
We want to check in the BL module - and possibly in the storage module as well...  What happens in the case where fundId is empty?  I suppose it means that toBudget will be empty.   Does this mean that subsequent queries against that table will fail, or just be empty?  

If they fail, that's fine - the transaction should then fail and an appropriate error response returned.  No need to explicitly check for toFund to exist in the storage module.  

If this doesn't cause a failure that's an issue and we should probably add an explicit check to make sure toFund exists in the storage module.

Reply:
In theory checking in the BL module should be sufficient since the storage module is intended for internal use only and clients would always work through the BL module.  What I describe above is just being extra careful, which can't hurt; especially when it comes to financial transactions

Reply:
Thanks for the clarification. I've decided to create an extra check for the destination budget existence. And I've added a unit test to exercise this case as well.